### PR TITLE
Llvm updates

### PR DESCRIPTION
**Pull request by aw1621107**
Add in some extra options for libcxx + libcxxabi, lld, polly, and sanitizers. Also fix some strangeness with code signing and building lldb.

See #44250 for previous discussion.

### CODE REVIEW

File: Library/Formula/llvm.rb [hunk 1]
```diff
@@ -30,11 +30,21 @@ class Llvm < Formula
       sha256 "ae9180466a23acb426d12444d866b266ff2289b266064d362462e44f8d4699f3"
     end
 
+    resource "extra-tools" do
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Please rename this resource back to `clang-extra-tools` and move it to where it used to be. It's easier to understand for everyone if we stick to the names used upstream and I see no good reason to rename/move it.

File: Library/Formula/llvm.rb [hunk 2]
```diff
@@ -66,10 +81,18 @@ class Llvm < Formula
       url "http://llvm.org/git/clang.git"
     end
 
+    resource "extra-tools" do
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Please do the same like for the stable resource, i.e. rename and move it back.

File: Library/Formula/llvm.rb [hunk 3]
```diff
@@ -78,17 +101,25 @@ class Llvm < Formula
...
+
+    resource "sanitizers" do
+      url "http://llvm.org/git/compiler-rt.git"
     end
   end
 
   option :universal
-  option "with-clang", "Build Clang support library"
+  option "with-clang", "Build the Clang compiler and support libraries"
+  option "with-extra-tools", "Build extra tools for Clang"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Let's stick with `with-clang-extra-tools` to be consistent with upstream and the resource name.

File: Library/Formula/llvm.rb [hunk 4]
```diff
@@ -106,6 +137,10 @@ class Llvm < Formula
     depends_on CodesignRequirement
   end
 
+  if build.with? "polly"
+    depends_on "isl"
+  end
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: You can shorten this to `depends_on "isl" if build.with? "polly"`.

File: Library/Formula/llvm.rb [hunk 5]
```diff
@@ -116,18 +151,74 @@ def install
...
+    if build.with? "clang"
+      (buildpath/"tools/clang").install resource("clang")
     end
 
-    if build.with? "clang"
+    if build.with? "extra-tools"
+      if build.with? "clang"
+        (buildpath/"tools/clang/tools/extra").install resource("extra-tools")
+      else
+        odie "clang's extra tools must be built with clang"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Please use `raise` instead of `odie` here (and everywhere else). Please also be consistent with capitalization: When referring to the compiler in text, it should be `Clang` (like in the option descriptions).

File: Library/Formula/llvm.rb [hunk 6]
```diff
@@ -116,18 +151,74 @@ def install
...
+      end
+    end
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: You can simplify this block to be shorter and reduce nesting like that:

```ruby
if build.with? "clang-extra-tools"
  raise "Clang's extra tools must be built with Clang" if build.without? "clang"
  (buildpath/"tools/clang/tools/extra").install resource("clang-extra-tools")
end
```

Please apply this idea also to similar blocks below.

File: Library/Formula/llvm.rb [hunk 7]
```diff
@@ -116,18 +151,74 @@ def install
...
+    if build.with? "libcxx"
       (buildpath/"projects/libcxx").install resource("libcxx")
-      (buildpath/"tools/clang").install resource("clang")
-      (buildpath/"tools/clang/tools/extra").install resource("clang-tools-extra")
+      (buildpath/"projects/libcxxabi").install resource("libcxxabi")
+    end
+
+    if build.with? "lld"
+      (buildpath/"tools/lld").install resource("lld")
+    end
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Shorten to `(buildpath/"tools/lld").install resource("lld") if build.with? "lld"` please. This also applies to similar constructs as long as such they comfortably fit on a single line.

File: Library/Formula/llvm.rb [hunk 8]
```diff
@@ -45,9 +55,14 @@ class Llvm < Formula
...
-    resource "clang-tools-extra" do
-      url "http://llvm.org/releases/3.6.2/clang-tools-extra-3.6.2.src.tar.xz"
-      sha256 "6a0ec627d398f501ddf347060f7a2ccea4802b2494f1d4fd7bda3e0442d04feb"
+    resource "polly" do
+      url "http://llvm.org/releases/3.6.2/polly-3.6.2.src.tar.xz"
+      sha256 "f2a956730b76212f22a1c10f35f195795e4d027ad28c226f97ddb8c0fd16bcbc"
+    end
+
+    resource "sanitizers" do
+      url "http://llvm.org/releases/3.6.2/compiler-rt-3.6.2.src.tar.xz"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Why is the resource named `sanitizers` when the file name in the URL starts with `compiler-rt`?

File: Library/Formula/llvm.rb [hunk 9]
```diff
@@ -116,18 +151,74 @@ def install
...
+      else
+        odie "sanitizers must be built along with clang"
+      end
     end
 
-    (buildpath/"tools/lld").install resource("lld") if build.with? "lld"
-    (buildpath/"tools/lldb").install resource("lldb") if build.with? "lldb"
+    if build.with?("tests") &&
+       build.with?("sanitizers") &&
+       build.without?("libcxx")
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: This looks short enough to comfortably fit on a single line. Also, have you forgotten to define the `--with-tests` option further up?

**Comment by UniqMartin** [comment]
I have added a bunch of remarks, but almost all of them are about style only. Please wait and see if the CI build succeeds before addressing them and pushing your updated branch. Thanks!

File: Library/Formula/llvm.rb (hunk 5 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: Hmmm, could have sworn someone else told me to use `odie` instead of `raise`, but I can't find the comment...

File: Library/Formula/llvm.rb (hunk 9 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: I think I had it earlier, but IIRC I took it out since it failed every time I tested it. Looks like I forgot to take the rest out when I decided not to include tests.

File: Library/Formula/llvm.rb (hunk 8 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: Thought the name described its contents better than `compiler-rt`, especially for the user-facing flags. That's assuming the sanitizers are the most widely used portion of the runtime libraries, though, so I should fix...

File: Library/Formula/llvm.rb (hunk 1 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: IIRC I moved it so the resources are in alphabetical order. Do you still want it under the `lldb` resource?

File: Library/Formula/llvm.rb (hunk 7 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: I think I used the nesting to avoid going over 80 characters. If over 80 characters still fits comfortably, I can change that.

File: Library/Formula/llvm.rb (hunk 9 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: And the condition is 85 characters on a single line, if that's fine

File: Library/Formula/llvm.rb (hunk 5 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: Yeah, looks like it got clobbered by my rebase/merge messup. Mike McQuaid told me to use `odie` for the libcxx sanitizer test check, and I took that to mean that all `raise`s should be replaced.

File: Library/Formula/llvm.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: I'm fine with it staying here to keep them in alphabetical order. Just wanted to avoid moving stuff around for no reason.

File: Library/Formula/llvm.rb (hunk 5 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: You can leave it as-is. Sorry for creating confusion! We haven't agreed yet on whether `odie` or `raise` is preferable, as those cases are relatively rare. (We'll fix that ourselves later, if necessary.)

File: Library/Formula/llvm.rb (hunk 7 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: In this particular instance it won't even hit 80 characters, so please change it.

File: Library/Formula/llvm.rb (hunk 8 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: It's fine to call the user-facing flag `--with-sanitizers`, but the resource should really be named `compiler-rt`. To make the connection between the two clearer, I'd suggest to add a code comment inside the `install` method where both the flag is evaluated and the resource is used.

File: Library/Formula/llvm.rb (hunk 9 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Keeping it below 80 characters for readability is fine. And if it's not used and not working, just remove it. Having working tests for such a complex piece of software would be really good, but maybe it's better to work on that in a separate PR.

File: Library/Formula/llvm.rb (hunk 5 above)
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: > Mike McQuaid told me to use `odie` for the libcxx sanitizer test check, and I took that to mean that all `raise`s should be replaced.

It would have been helpful if you referenced the previous discussion. I edited your initial comment for the benefit of others.

**Comment by aw1621107** [comment]
Alright, updated. Hope this addresses any issues.

`brew audit` complains

    python modules have explicit framework links

Is there anything I can do about that? I'm not familiar with the way the Python side of things work...

**Comment by tdsmith** [comment]
@UniqMartin asked me to comment on the Python issue; Jenkins suggests that the default build passes `brew audit` so I'm guessing your local version is installed with non-default options and that this is probably the same lldb-related issue mentioned in #45787. IMHO I don't think it's a showstopper.

File: Library/Formula/llvm.rb [hunk 10]
```diff
@@ -45,9 +55,14 @@ class Llvm < Formula
...
 
-    resource "clang-tools-extra" do
-      url "http://llvm.org/releases/3.6.2/clang-tools-extra-3.6.2.src.tar.xz"
-      sha256 "6a0ec627d398f501ddf347060f7a2ccea4802b2494f1d4fd7bda3e0442d04feb"
+    resource "polly" do
+      url "http://llvm.org/releases/3.6.2/polly-3.6.2.src.tar.xz"
+      sha256 "f2a956730b76212f22a1c10f35f195795e4d027ad28c226f97ddb8c0fd16bcbc"
+    end
+
+    resource "compiler-rt" do
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Just a detail, but now that it's called `compiler-rt` you probably want to move it up to keep the alphabetical order.

File: Library/Formula/llvm.rb [hunk 11]
```diff
@@ -78,17 +101,27 @@ class Llvm < Formula
       url "http://llvm.org/git/lldb.git"
     end
 
-    resource "clang-tools-extra" do
-      url "http://llvm.org/git/clang-tools-extra.git"
+    resource "polly" do
+      url "http://llvm.org/git/polly.git"
+    end
+
+    resource "compiler-rt" do
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Just a detail, but now that it's called `compiler-rt` you probably want to move it up to keep the alphabetical order.

File: Library/Formula/llvm.rb [hunk 12]
```diff
@@ -78,17 +101,27 @@ class Llvm < Formula
...
+
   option :universal
-  option "with-clang", "Build Clang support library"
+  option "with-clang", "Build the Clang compiler and support libraries"
+  option "with-clang-extra-tools", "Build extra tools for Clang"
+  option "with-libcxx", "Build the libc++ standard library"
   option "with-lld", "Build LLD linker"
   option "with-lldb", "Build LLDB debugger"
   option "with-rtti", "Build with C++ RTTI"
+  option "with-polly", "Build with the experimental Polly optimizer"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: You might want to put that between `with-lldb` and `with-rtti` to keep the options sorted.

File: Library/Formula/llvm.rb [hunk 13]
```diff
@@ -78,17 +101,27 @@ class Llvm < Formula
...
   option "with-python", "Build Python bindings against Homebrew Python"
+  option "with-compiler-rt", "Build Clang runtime support libraries for code sanitizers, builtins, and profiling"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: And this one could be placed after `with-clang-extra-tools` …

File: Library/Formula/llvm.rb [hunk 14]
```diff
@@ -116,18 +149,49 @@ def install
...
-      raise "Building LLDB needs Clang support library."
-    end
+    (buildpath/"tools/clang").install resource("clang") if build.with? "clang"
 
-    if build.with? "clang"
-      (buildpath/"projects/libcxx").install resource("libcxx")
-      (buildpath/"tools/clang").install resource("clang")
-      (buildpath/"tools/clang/tools/extra").install resource("clang-tools-extra")
+    if build.with? "clang-extra-tools"
+      odie "clang's extra tools must be built with clang" if build.without? "clang"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: I would prefer if you could refer to the various components as `Clang`, `LLDB`, etc. like in the option descriptions. (This also applies to all other `odie` statements.) Alternatively, you could write something like `--with-clang-extra-tools requires --with-clang`, which is somewhat more technical, but has the advantage of being very clear (and a user can copy & paste the option to retry the installation).

File: Library/Formula/llvm.rb [hunk 15]
```diff
@@ -116,18 +149,49 @@ def install
...
+      (buildpath/"tools/clang/tools/extra").install resource("clang-extra-tools")
     end
 
+    (buildpath/"projects/libcxx").install resource("libcxx") if build.with? "libcxx"
+    (buildpath/"projects/libcxxabi").install resource("libcxxabi") if build.with? "libcxx"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Because both lines use the same condition, it is better to use an `if` block here. (But thanks for reviewing and updating all the other instances where a trailing `if` modifier is better!)

**Comment by UniqMartin** [comment]
Thanks for addressing all my comments! I have added a few more code remarks, but nothing major.

I tried to build the formula locally (OS X 10.11.2, Xcode 7.2). The default build (with no options) and a build with `--with-clang --with-clang-extra-tools --with-compiler-rt --with-libcxx --with-lld --with-lldb` succeeded. But when I add `--with-polly`, the build fails with errors related to `isl`. Have you tried the `--with-polly` build yourself? Could it be related to the recent update of `isl` to 0.15? (We could remove this part, if the issue cannot be easily resolved. You can still add it later in a new PR.)

Other than the `--with-polly` issue, things are looking pretty good. Can you squash all those individual commits into one commit and describe briefly in the commit message what the important changes are? You can also post the summary here, if you are not entirely comfortable doing the squashing yourself. (I'll add it to the squashed commit when I merge.)

File: Library/Formula/llvm.rb (hunk 10 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: Whoops, forgot about that...

**Comment by aw1621107** [comment]
Alright, hopefully fixed the rest of the things.

Oddly enough, a `--HEAD` build with `--with-clang --with-libcxx --with-lld --with-lldb --with-polly --with-python` and `isl` 0.15 worked fine. Maybe it's something that was fixed in a newer version of Polly?

And that Python issue looks like it's going to be fun to fix...

**Comment by UniqMartin** [comment]
> Alright, hopefully fixed the rest of the things.

Did you forget to push your changes?

> Oddly enough, a `--HEAD` build with `--with-clang --with-libcxx --with-lld --with-lldb --with-polly --with-python` and `isl` 0.15 worked fine. Maybe it's something that was fixed in a newer version of Polly?

The explanation for this is probably (from the [LLVM 3.7 release notes](http://llvm.org/releases/3.7.0/docs/ReleaseNotes.html#polly-the-polyhedral-loop-optimizer-in-llvm)): “isl, the math library Polly uses, has been imported into the source code repository of Polly and is now distributed as part of Polly.” My suggestion would be to drop everything related to Polly from this PR, so that we can merge the rest soon and other Homebrew users can start benefiting from your work. (The Polly issue could be addressed afterwards in a new PR.) What do you think?

> I'm on OS X 10.11.1, Xcode 7.1. Hopefully it's not something in the update that broke things...

I don't think this is the issue. As far as I can tell, the differences between Xcode 7.1 and 7.2 are minimal on the compiler level.

> And that Python issue looks like it's going to be fun to fix...

We don't have to fix it in this PR or in Homebrew. People can still switch between a system Python and any other Python they have in PATH by either providing or omitting `--with-python` (if the description is correct). If you feel very passionate about this issue, I'd suggest to get in contact with LLVM upstream. If it can be fixed there, everyone will benefit and we won't need to hack around this issue. (Have a look at #45787 and [this upstream bug report](https://llvm.org/bugs/show_bug.cgi?id=25446) if you are interested and haven't seen them yet.)

**Comment by aw1621107** [comment]
Yeah, looked like I forgot to push...

Perhaps Polly could be `--HEAD-only` while the formula is still on 3.6? I'm fine with removing it otherwise.

If I find time for the Python thing, I'll take a look.

**Comment by UniqMartin** [comment]
> Perhaps Polly could be `--HEAD-only` while the formula is still on 3.6? I'm fine with removing it otherwise.

Yes, you are right. Let's move the `--with-polly` option and the `isl` dependency into the `head do` block and please add a short code comment about why it only applied to `HEAD`. Otherwise, the formula is looking good! :+1:

Also I'm quoting myself here from [a previous comment](https://github.com/Homebrew/homebrew/pull/46721#issuecomment-163139374), just in case you overlooked it:

> Can you squash all those individual commits into one commit and describe briefly in the commit message what the important changes are? You can also post the summary here, if you are not entirely comfortable doing the squashing yourself.

How would you like to handle this? Either way is fine with me, just let me know.

**Comment by aw1621107** [comment]
Whoops, totally forgot about the squash. Wouldn't that involve a force push/new PR?

**Comment by aw1621107** [comment]
Would it be possible for LLVM stable to depend on isl014 to install Polly? Or can dependencies only be from the main homebrew library?

**Comment by aw1621107** [comment]
Scratch that, looks like it won't work.

Right now I'm failing the build if `--with-polly` is provided without `--HEAD` by using the following:

    odie "--with-polly requires --HEAD" unless build.head?

Is there a more idiomatic way of doing this?

**Comment by UniqMartin** [comment]
> Whoops, totally forgot about the squash. Wouldn't that involve a force push/new PR?

Yes, that would involve a force push to your `llvm-update` branch that is associated with this PR. (A new PR is not necessary.) But as I said, if you fear you'll mess up your branch by doing it yourself, feel free to post a summary that would go into the squash commit and I'll handle the rest. (Or experiment in a separate Git repository with a dummy project on GitHub, if you want to learn how to do that and gain some confidence in doing that.)

> Would it be possible for LLVM stable to depend on `isl014` to install Polly? Or can dependencies only be from the main homebrew library?

In theory, this is possible as long as the non-core dependency is strictly optional. (This would be the case.) In practice, there are some technical limitations that currently make this impossible, so we better don't follow that path right now.

> Right now I'm failing the build if `--with-polly` is provided without `--HEAD` by using the following: […] Is there a more idiomatic way of doing this?

You don't have to perform this check at all. Let's make `--with-polly` a HEAD-only option for now. This entails moving the corresponding `option` to the end of the `head do` block and removing the `polly` resource in the `stable do` block. And as the `isl` dependency is already vendored into Polly, you can remove the `depends_on "isl"` line completely (sorry, I wrote something different about that in my previous comment, because I momentarily forgot about `isl` being vendored into Polly).

File: Library/Formula/llvm.rb [hunk 16]
```diff
@@ -45,9 +60,9 @@ class Llvm < Formula
       sha256 "940dc96b64919b7dbf32c37e0e1d1fc88cc18e1d4b3acf1e7dfe5a46eb6523a9"
     end
 
-    resource "clang-tools-extra" do
-      url "http://llvm.org/releases/3.6.2/clang-tools-extra-3.6.2.src.tar.xz"
-      sha256 "6a0ec627d398f501ddf347060f7a2ccea4802b2494f1d4fd7bda3e0442d04feb"
+    resource "polly" do
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Just to complement [my comment](https://github.com/Homebrew/homebrew/pull/46721#issuecomment-164157454): Remove this resource.

File: Library/Formula/llvm.rb [hunk 17]
```diff
@@ -78,17 +105,23 @@ class Llvm < Formula
...
   option :universal
-  option "with-clang", "Build Clang support library"
+  option "with-clang", "Build the Clang compiler and support libraries"
+  option "with-clang-extra-tools", "Build extra tools for Clang"
+  option "with-compiler-rt", "Build Clang runtime support libraries for code sanitizers, builtins, and profiling"
+  option "with-libcxx", "Build the libc++ standard library"
   option "with-lld", "Build LLD linker"
   option "with-lldb", "Build LLDB debugger"
-  option "with-rtti", "Build with C++ RTTI"
+  option "with-polly", "Build with the experimental Polly optimizer"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Just to complement [my comment](https://github.com/Homebrew/homebrew/pull/46721#issuecomment-164157454): Move this option at the end of the `head do` block and add a short comment why it only applies to HEAD (like we've discussed).

File: Library/Formula/llvm.rb [hunk 18]
```diff
@@ -106,7 +139,7 @@ class Llvm < Formula
     depends_on CodesignRequirement
   end
 
-  keg_only :provided_by_osx
+  depends_on "isl" if build.with? "polly"
```
Comment (UniqMartin) [inline Library/Formula/llvm.rb]: Just to complement [my comment](https://github.com/Homebrew/homebrew/pull/46721#issuecomment-164157454): Remove this dependency.

File: Library/Formula/llvm.rb (hunk 17 above)
Comment (aw1621107) [inline Library/Formula/llvm.rb]: Ohhh, so *that*'s what you're supposed to do. I misinterpreted your comment and just got rid of the stable version of the resource, leaving the option where it was. I'll fix that...

**Comment by aw1621107** [comment]
Alright, pushed the changes.

Moved `--with-polly` right underneath the `"polly"` resource in the `head do` block, and it's no longer showing up in `brew info llvm`. Is this expected?

Also, executing `brew install llvm --with-polly` starts installing the stable version as if the flag weren't given. I thought `--HEAD`-only options would force a `--HEAD` build?

I think I can try squashing myself. I'll wait until everything's cleaned up though

**Comment by UniqMartin** [comment]
> Alright, pushed the changes.

Changes are looking good. :+1:

> Moved `--with-polly` right underneath the `"polly"` resource in the `head do` block, and it's no longer showing up in `brew info llvm`. Is this expected?

Yes, unfortunately `brew info llvm` can only show the default/stable options. The `--with-polly` option can be discovered via `brew options llvm --HEAD`. (We know that's not very intuitive, but that will hopefully be fixed in the future.)

> Also, executing `brew install llvm --with-polly` starts installing the stable version as if the flag weren't given. I thought `--HEAD`-only options would force a `--HEAD` build?

Unfortunately, `brew install` (like most `brew` commands) accepts arbitrary options and simply ignores the ones it doesn't know about. Thus, there's no diagnostic and one has to additionally pass `--HEAD` to get the desired effect. (That's another issue we know about, but that hasn't been addressed yet.)

**Comment by aw1621107** [comment]
Alright, squash completed!

Maybe I'll take a look at those flag things over break...

**Comment by UniqMartin** [comment]
@BrewTestBot test this please

**Comment by UniqMartin** [comment]
> Alright, squash completed!

Merged in 7a72f5f6b02e9bf996cf7b146fe601b2916ad223. Thank you for this contribution to Homebrew and the nontrivial amount of work you've put into improving this formula @aw1621107! :tada:

> Maybe I'll take a look at those flag things over break...

That would be awesome!

**Comment by aw1621107** [comment]
Thank you for your help!
